### PR TITLE
Create SingleSpeedTalonDriveSubsystem

### PR DESCRIPTION
**Pull request by Stick-2357**
Might need to add more configuration to the talons, and need to set defaultCommand. Will need to make the command before we can do that.

### CODE REVIEW

File: src/main/java/com/systemmeltdown/robotlib/subsystems/drive/SingleSpeedTalonDriveSubsystem.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+        }
+
+        // TODO: add more config to talons?
+    }
+
+    @Override
+    protected int getCurrentSpeedLeftClicksPerSecond() {
+        double proportion = m_leftMaster.get();
+        double inches = proportion * getMaxSpeedInchesPerSecond();
+        double clicks = inches * getClicksPerInch();
```
Comment (coderkevin) [inline src/main/java/com/systemmeltdown/robotlib/subsystems/drive/SingleSpeedTalonDriveSubsystem.java]: We should not be estimating clicks based on proportion. We should be getting actual click readings from the encoders connected to the Talon SRX.

File: src/main/java/com/systemmeltdown/robotlib/subsystems/drive/SingleSpeedTalonDriveSubsystem.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+        return (int) clicks;
+    }
+
+    @Override
+    protected int getCurrentSpeedRightClicksPerSecond() {
+        double proportion = m_rightMaster.get();
+        double inches = proportion * getMaxSpeedInchesPerSecond();
+        double clicks = inches * getClicksPerInch();
```
Comment (coderkevin) [inline src/main/java/com/systemmeltdown/robotlib/subsystems/drive/SingleSpeedTalonDriveSubsystem.java]: Same here, get real sensor values, don't fake them.

File: src/main/java/com/systemmeltdown/robotlib/subsystems/drive/SingleSpeedTalonDriveSubsystem.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+        m_rightMaster.set(ControlMode.PercentOutput, rightProportion);
+    }
+
+    @Override
+    protected void setVelocity(int leftClicksPerSecond, double rightClicksPerSecond) {
+        double leftClicksPer100ms = leftClicksPerSecond / 10;
+        double rightClicksPer100ms = rightClicksPerSecond / 10;
+
+        m_leftMaster.set(ControlMode.Velocity, leftClicksPer100ms);
+        m_rightMaster.set(ControlMode.Velocity, rightClicksPer100ms);
```
Comment (coderkevin) [inline src/main/java/com/systemmeltdown/robotlib/subsystems/drive/SingleSpeedTalonDriveSubsystem.java]: We should be setting a Velocity PID value here, not estimating a proportion. If we're not ready for that in this PR, we can leave a TODO comment instead. Let's not fake behavior here.

File: src/main/java/com/systemmeltdown/robotlib/subsystems/drive/SkidSteerDriveSubsystem.java [hunk 4]
```diff
@@ -37,6 +37,10 @@
 	private int m_clicksPerInch = 0;
 	private int m_maxSpeedClicksPerSecond = 0;
 
+	public int getClicksPerInch() {
```
Comment (coderkevin) [inline src/main/java/com/systemmeltdown/robotlib/subsystems/drive/SkidSteerDriveSubsystem.java]: Let's not expose any public methods based on clicks. Let's use real measurable values for our public interfaces.

File: src/main/java/com/systemmeltdown/robotlib/subsystems/drive/SkidSteerDriveSubsystem.java [hunk 5]
```diff
@@ -6,30 +6,27 @@
 import com.systemmeltdown.robotlib.util.RobotMath;
 
 /**
- * Base class for any kind of "Skid Steer" drive base.
- * This makes assumptions that we will use encoders and velocity drive.
- * However, this makes zero assumptions about hardware or implementation of such.
+ * Base class for any kind of "Skid Steer" drive base. This makes assumptions
```
Comment (coderkevin) [inline src/main/java/com/systemmeltdown/robotlib/subsystems/drive/SkidSteerDriveSubsystem.java]: Why are these formatting changes included in this PR? We should not be doing widespread formatting in the same PR as real code changes.

File: src/main/java/com/systemmeltdown/robotlib/subsystems/drive/SingleSpeedTalonDriveSubsystem.java [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+
+public class SingleSpeedTalonDriveSubsystem extends SkidSteerDriveSubsystem {
+    private WPI_TalonSRX m_rightMaster;
+    private WPI_TalonSRX m_leftMaster;
+    private WPI_TalonSRX[] m_rightSlaves;
+    private WPI_TalonSRX[] m_leftSlaves;
+
+    public SingleSpeedTalonDriveSubsystem(int rightMasterId, int leftMasterId, int[] rightSlaveIds,
+            int[] leftSlaveIds) {
+
```
Comment (coderkevin) [inline src/main/java/com/systemmeltdown/robotlib/subsystems/drive/SingleSpeedTalonDriveSubsystem.java]: The configuration in this constructor looks fine for a first run, but I feel this code should be in an overridden `configure` method, and the first line in the configure should be to set the talons back to factory configuration.

See here for an example: https://github.com/CrossTheRoadElec/Phoenix-Examples-Languages/blob/master/Java/ArcadeDrive_ArbFeedForward/src/main/java/frc/robot/Robot.java

File: src/main/java/com/systemmeltdown/robotlib/controllers/DriverControls.java [hunk 7]
```diff
@@ -24,12 +24,14 @@ public DriverControls(XboxController controller, double deadband) {
   @Override
   public double getSpeed() {
     double value = m_controller.getY(Hand.kLeft);
+    System.out.println("Left value: " + value);
```
Comment (coderkevin) [inline src/main/java/com/systemmeltdown/robotlib/controllers/DriverControls.java]: Let's remove the printlns before merge.

File: src/main/java/com/systemmeltdown/robotlib/subsystems/drive/SingleSpeedTalonDriveSubsystem.java [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+public class SingleSpeedTalonDriveSubsystem extends SkidSteerDriveSubsystem {
+    private WPI_TalonSRX m_rightMaster;
+    private WPI_TalonSRX m_leftMaster;
+    private WPI_TalonSRX[] m_rightSlaves;
+    private WPI_TalonSRX[] m_leftSlaves;
+
+    public SingleSpeedTalonDriveSubsystem(
+        TalonIDGroup rightTalonIDs,
+        TalonIDGroup leftTalonIDs) {
```
Comment (coderkevin) [inline src/main/java/com/systemmeltdown/robotlib/subsystems/drive/SingleSpeedTalonDriveSubsystem.java]: I'm not sure having a class solely for the commuting of ids is better than just giving arrays of ids here.

I think if we have a small class to group talons, it should be more functional and manage the WPI_TalonSRX objects and their configs.

Also, I'd rather not have configuration options like inversion in the constructor. I'd rather pass those through the `configure()` method which is designed for this.

Comment (Stick-2357) [reply src/main/java/com/systemmeltdown/robotlib/subsystems/drive/SingleSpeedTalonDriveSubsystem.java]: When you mention the `configure()` method, I assume you're talking about the same one you mentioned overriding earlier. What object is the method a member of? For now, I've added a `configure(boolean isInverted)` method to the newly made TalonGroup.

Comment (coderkevin) [reply src/main/java/com/systemmeltdown/robotlib/subsystems/drive/SingleSpeedTalonDriveSubsystem.java]: This configure in SkidSteerDriveSubsystem:
https://github.com/frc2357/robotlib2357/blob/5ebd2ca62e1007a9337ccca889c899ae6c8a553a/src/main/java/com/systemmeltdown/robotlib/subsystems/drive/SkidSteerDriveSubsystem.java#L43

It's designed to be overridden, pluck out the settings from the map you care about, then pass on the map to the superclass.

Comment (coderkevin) [reply src/main/java/com/systemmeltdown/robotlib/subsystems/drive/SingleSpeedTalonDriveSubsystem.java]: I would suggest adding configuration values to the map like `"invertLeft"` = `true` and `"invertRight"` = `true`

The real advantage of this is you can keep all these configuration options together in one place and in one function call.

In the future, we'll add more configuration options to this map, like specific ones for the TalonSRX for ramping, PIDF values, etc.

File: src/main/java/com/systemmeltdown/robotlib/subsystems/drive/SingleSpeedTalonDriveSubsystem.java [hunk 9]
```diff
@@ -1,62 +1,40 @@
...
+        // we don't have to call getMasterTalon
+        return m_rightTalonGroup.getMasterTalon().getSelectedSensorPosition();
     }
 
     @Override
     protected void setProportional(double leftProportion, double rightProportion) {
-        m_leftMaster.set(ControlMode.PercentOutput, leftProportion);
-        m_rightMaster.set(ControlMode.PercentOutput, rightProportion);
+        // could add a set command to the TalonGroup, so that
+        // we don't have to call getMasterTalon
```
Comment (coderkevin) [inline src/main/java/com/systemmeltdown/robotlib/subsystems/drive/SingleSpeedTalonDriveSubsystem.java]: I would recommend putting what you can in the TalonGroup, for sure.

Comment (Stick-2357) [reply src/main/java/com/systemmeltdown/robotlib/subsystems/drive/SingleSpeedTalonDriveSubsystem.java]: Oh whoops, I ended up going ahead and doing that, just forgot to remove the comment.

File: src/main/java/com/systemmeltdown/robotlib/subsystems/pdp/PDPSubsystem.java [hunk 10]
```diff
@@ -0,0 +1,64 @@
+package com.systemmeltdown.robotlib.subsystems.pdp;
```
Comment (coderkevin) [inline src/main/java/com/systemmeltdown/robotlib/subsystems/pdp/PDPSubsystem.java]: You need to rebase this PR onto current master. You shouldn't have this here.

File: src/main/java/com/systemmeltdown/robotlib/subsystems/drive/SingleSpeedTalonDriveSubsystem.java [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+public class SingleSpeedTalonDriveSubsystem extends SkidSteerDriveSubsystem {
+    /**
+	 * Whether or not the left talon group needs to be inverted Value: boolean
+	 */
+    public static final String CONFIG_IS_LEFT_INVERTED = "is_left_inverted";
+    
+    /**
+	 * Whether or not the right talon group needs to be inverted Value: boolean
+	 */
+	public static final String CONFIG_IS_RIGHT_INVERTED = "is_right_inverted";
```
Comment (CrazyFloob) [inline src/main/java/com/systemmeltdown/robotlib/subsystems/drive/SingleSpeedTalonDriveSubsystem.java]: Indented to far.

**Pull request closed without merging**